lib/puppet/util/windows/service.rb

```diff
@@ -596,75 +680,109 @@ def query_config(service)
 
       # @api private
-      # waits for a windows service to report final_state if it
-      # is in pending_state
+      # Waits for a service to finish transitioning from
+      # a pending state. The service must be in the pending state
+      # before invoking this routine.
       #
       # @param [:handle] service handle to the service to wait on
-      # @param [Integer] pending_states array of acceptable states to wait on
-      # @param [Integer] final_state the state indicating the transition is finished
-      # @return [bool] 'true' once the service is reporting final_state,
-      #   'false' if the service never reached final_state or was not in pending_states
-      def wait_on_pending_transition(service, pending_state, final_state, raise_on_timeout: false)
+      # @param [Integer] pending_state the pending state
+      def wait_on_pending_state(service, pending_state)
+        final_state = FINAL_STATES[pending_state]
+
+        Puppet.debug _("Waiting for the pending transition to the %{final_state} state to finish.") % { final_state: SERVICE_STATES[final_state] }
+
         elapsed_time = 0
         last_checkpoint = -1
         loop do
           status = query_status(service)
           state = status[:dwCurrentState]
-          return true if state == final_state
+
+          # Check if our service has finished transitioning to
+          # the final_state OR if an unexpected transition
+          # has occurred
+          return if state == final_state
           unless state == pending_state
-            if raise_on_timeout
-              raise Puppet::Error.new(_("Service was not in pending state: %{pending_state}, current state is %{current_state}") % { pending_state: SERVICE_STATES[pending_state], current_state: SERVICE_STATES[state] })
-            else
-              return false
-            end
+            raise Puppet::Error, _("Unexpected transition to the %{current_state} state while waiting for the pending transition from %{pending_state} to %{final_state} to finish.") % { current_state: SERVICE_STATES[state], pending_state: SERVICE_STATES[pending_state], final_state: SERVICE_STATES[final_state] }
           end
-          # When the service is in the pending state we need to do the following:
-          # 1. check if any progress has been made since dwWaitHint using dwCheckPoint,
-          #    and fail if no progress was made
-          # 2. if progress has been made, increment elapsed_time and set last_checkpoint
-          # 3. sleep, then loop again if there was progress.
-          time_to_wait = wait_hint_to_wait_time(status[:dwWaitHint])
+
+          # Check if any progress has been made since our last sleep
+          # using the dwCheckPoint. If no progress has been made then
+          # check if we've timed out, and raise an error if so
           if status[:dwCheckPoint] > last_checkpoint
             elapsed_time = 0
+            last_checkpoint = status[:dwCheckPoint]
```

Comment:
Since we set elapsed_time to 0 in line 750 and dont modify it before line 768, do we need both?

Reply:
Yes. Line 750 sets it to the scope outside of the begin block. Otherwise, the elapsed_time variable will only be scoped to the begin end block which means it cannot be referenced in latter iterations of the loop.

Reply:
Ya. I get that. I was wondering why it needed to be (re)set inside the block when there is nothing that changes it in-between the two referenced locations.

Reply:
Aha. I get it now. Nvm.